{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m22.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: numpy in /home/rickbook/.local/lib/python3.10/site-packages (from biopython) (1.24.2)\n",
      "Installing collected packages: biopython\n",
      "Successfully installed biopython-1.81\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio import SeqIO, Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simple way to retrive protein sequence\n",
    "\n",
    "#Setting up your email to be able to use Entrez\n",
    "Entrez.email = 'riccardo.tedoldi@studenti.unitn.it'\n",
    "\n",
    "#Here, we set up a temporary handle with our downloaded sequence in fasta format\n",
    "temp = Entrez.efetch(db=\"protein\",rettype=\"fasta\",id=\"2KIU_A\")\n",
    "\n",
    "#Reading the sequence information as a string in fasta format\n",
    "aaseq = SeqIO.read(temp, format=\"fasta\")\n",
    "\n",
    "#Creating a fasta file to write our downloaded sequence\n",
    "aaseq_out = open(\"2KIU_A.fasta\",'w')\n",
    "\n",
    "#Writing the sequence record in fasta format\n",
    "SeqIO.write(aaseq,aaseq_out,\"fasta\")\n",
    "\n",
    "#Closing both the temp handle and the FASTA file\n",
    "temp.close()\n",
    "aaseq_out.close()\n",
    "\n",
    "# from Bio.Seq import Seq\n",
    "# my_seq = Seq(\"AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTVDEVEFQKRRPQK\")\n",
    "\n",
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"pdb\", aaseq.seq, entrez_query='human[organism]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_record = NCBIXML.read(result_handle)\n",
    "\n",
    "result_handle.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****Alignment****\n",
      "sequence: pdb|2KIU|A Chain A, Forkhead box protein P1 [Homo sapiens]\n",
      "length: 87\n",
      "e value: 2.07572e-61\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTV...\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTV...\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTV...\n",
      "****Alignment****\n",
      "sequence: pdb|6XAT|A Chain A, FOXP4 protein [Homo sapiens]\n",
      "length: 110\n",
      "e value: 1.10651e-54\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTV...\n",
      "A+VRPPFTYASLIRQAILE+P++QLTLNEIYNWFTRMF YFRRN ATWKNAVRHNLSLHK FVRVENVKGAVWTV...\n",
      "ADVRPPFTYASLIRQAILETPDRQLTLNEIYNWFTRMFAYFRRNTATWKNAVRHNLSLHKCFVRVENVKGAVWTV...\n",
      "****Alignment****\n",
      "sequence: pdb|2A07|F Chain F, Forkhead box protein P2 [Homo sapiens] >pdb|2A07|G Chain G, Forkhead box protein P2 [Homo sapiens] >pdb|2A07|H Chain H, Forkhead box protein P2 [Homo sapiens] >pdb|2A07|I Chain I, Forkhead box protein P2 [Homo sapiens] >pdb|2A07|J Chain J, Forkhead box protein P2 [Homo sapiens] >pdb|2A07|K Chain K, Forkhead box protein P2 [Homo sapiens] >pdb|2AS5|F Chain F, Forkhead box protein P2 [Homo sapiens] >pdb|2AS5|G Chain G, Forkhead box protein P2 [Homo sapiens]\n",
      "length: 93\n",
      "e value: 1.38721e-52\n",
      "VRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTVDE...\n",
      "VRPPFTYA+LIRQAI+ES ++QLTLNEIY+WFTR F YFRRNAATWKNAVRHNLSLHK FVRVENVKGAVWTVDE...\n",
      "VRPPFTYATLIRQAIMESSDRQLTLNEIYSWFTRTFAYFRRNAATWKNAVRHNLSLHKCFVRVENVKGAVWTVDE...\n",
      "****Alignment****\n",
      "sequence: pdb|3QRF|F Structure of a domain-swapped FOXP3 dimer [Homo sapiens] >pdb|3QRF|G Structure of a domain-swapped FOXP3 dimer [Homo sapiens] >pdb|3QRF|H Structure of a domain-swapped FOXP3 dimer [Homo sapiens] >pdb|3QRF|I Structure of a domain-swapped FOXP3 dimer [Homo sapiens] >pdb|4WK8|F FOXP3 forms a domain-swapped dimer to bridge DNA [Homo sapiens] >pdb|4WK8|G FOXP3 forms a domain-swapped dimer to bridge DNA [Homo sapiens]\n",
      "length: 82\n",
      "e value: 3.41181e-44\n",
      "VRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENVKGAVWTVDE...\n",
      "+RPPFTYA+LIR AILE+PEKQ TLNEIY+WFTRMF +FR + ATWKNA+RHNLSLHK FVRVE+ KGAVWTVDE...\n",
      "MRPPFTYATLIRWAILEAPEKQRTLNEIYHWFTRMFAFFRNHPATWKNAIRHNLSLHKCFVRVESEKGAVWTVDE...\n",
      "****Alignment****\n",
      "sequence: pdb|2C6Y|A Crystal structure of interleukin enhancer-binding factor 1 bound to DNA [Homo sapiens] >pdb|2C6Y|B Crystal structure of interleukin enhancer-binding factor 1 bound to DNA [Homo sapiens]\n",
      "length: 111\n",
      "e value: 5.85515e-26\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKG...\n",
      "++ +PP++YA LI QAI  +P+KQLTLN IY   T+ +PY+R     W+N++RHNLSL++YF++V     E  KG...\n",
      "SDSKPPYSYAQLIVQAITMAPDKQLTLNGIYTHITKNYPYYRTADKGWQNSIRHNLSLNRYFIKVPRSQEEPGKG...\n",
      "****Alignment****\n",
      "sequence: pdb|1JXS|A Chain A, interleukin enhancer binding factor [Homo sapiens]\n",
      "length: 98\n",
      "e value: 1.84911e-25\n",
      "EVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGA...\n",
      "+ +PP++YA LI QAI  +P+KQLTLN IY   T+ +PY+R     W+N++RHNLSL++YF++V     E  KG+...\n",
      "DSKPPYSYAQLIVQAITMAPDKQLTLNGIYTHITKNYPYYRTADKGWQNSIRHNLSLNRYFIKVPRSQEEPGKGS...\n",
      "****Alignment****\n",
      "sequence: pdb|7CBY|C Structure of FOXG1 DNA binding domain bound to DBE2 DNA site [Homo sapiens]\n",
      "length: 106\n",
      "e value: 5.87047e-23\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGAVW...\n",
      "+PPF+Y +LI  AI +SPEK+LTLN IY +  + FPY+R N   W+N++RHNLSL+K FV+V     +  KG  W...\n",
      "KPPFSYNALIMMAIRQSPEKRLTLNGIYEFIMKNFPYYRENKQGWQNSIRHNLSLNKCFVKVPRHYDDPGKGNYW...\n",
      "****Alignment****\n",
      "sequence: pdb|7VOX|A Chain A, Hepatocyte nuclear factor 3-alpha [Homo sapiens] >pdb|7VOX|B Chain B, Hepatocyte nuclear factor 3-alpha [Homo sapiens] >pdb|7VOX|C Chain C, Hepatocyte nuclear factor 3-alpha [Homo sapiens] >pdb|7VOX|H Chain H, Hepatocyte nuclear factor 3-alpha [Homo sapiens]\n",
      "length: 102\n",
      "e value: 6.64473e-22\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENV-----KG...\n",
      "+  +PP++Y SLI  AI ++P K LTL+EIY W   +FPY+R+N   W+N++RH+LS +  FV+V        KG...\n",
      "SHAKPPYSYISLITMAIQQAPSKMLTLSEIYQWIMDLFPYYRQNQQRWQNSIRHSLSFNDCFVKVARSPDKPGKG...\n",
      "****Alignment****\n",
      "sequence: pdb|1D5V|A Chain A, S12 TRANSCRIPTION FACTOR (FKH-14) [Homo sapiens]\n",
      "length: 94\n",
      "e value: 9.78245e-22\n",
      "VRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGAV...\n",
      "V+PP++Y +LI  AI  +PEK++TLN IY +    FP++R N   W+N++RHNLSL++ FV+V     +  KG+ ...\n",
      "VKPPYSYIALITMAIQNAPEKKITLNGIYQFIMDRFPFYRENKQGWQNSIRHNLSLNECFVKVPRDDKKPGKGSY...\n",
      "****Alignment****\n",
      "sequence: pdb|6O3T|A Structural basis of FOXC2 and DNA interactions [Homo sapiens] >pdb|6O3T|B Structural basis of FOXC2 and DNA interactions [Homo sapiens]\n",
      "length: 95\n",
      "e value: 1.27791e-21\n",
      "VRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGAV...\n",
      "V+PP++Y +LI  AI  +PEK++TLN IY +    FP++R N   W+N++RHNLSL++ FV+V     +  KG+ ...\n",
      "VKPPYSYIALITMAIQNAPEKKITLNGIYQFIMDRFPFYRENKQGWQNSIRHNLSLNECFVKVPRDDKKPGKGSY...\n",
      "****Alignment****\n",
      "sequence: pdb|6NCE|A Crystal structure of the human FOXN3 DNA binding domain in complex with a forkhead DNA sequence [Homo sapiens] >pdb|6NCM|A Crystal structure of the human FOXN3 DNA binding domain in complex with a forkhead-like (FHL) DNA sequence [Homo sapiens] >pdb|6NCM|B Crystal structure of the human FOXN3 DNA binding domain in complex with a forkhead-like (FHL) DNA sequence [Homo sapiens]\n",
      "length: 100\n",
      "e value: 2.38762e-21\n",
      "AEVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVEN------VK...\n",
      "   +PP++++ LI  AI +SP K+L + +IYNW    FPYF      WKN+VRHNLSL+K F +V+        K...\n",
      "GNCKPPYSFSCLIFMAIEDSPTKRLPVKDIYNWILEHFPYFANAPTGWKNSVRHNLSLNKCFKKVDKERSQSIGK...\n",
      "****Alignment****\n",
      "sequence: pdb|6AKO|C Crystal Structure of FOXC2 DBD Bound to DBE2 DNA [Homo sapiens] >pdb|6AKP|C Crystal Structural of FOXC2 DNA binding domain bound to PC promoter [Homo sapiens] >pdb|6LBM|C Chain C, Forkhead box protein C2 [Homo sapiens]\n",
      "length: 103\n",
      "e value: 4.56161e-21\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGAVW...\n",
      "+PP++Y +LI  AI  +PEK++TLN IY +    FP++R N   W+N++RHNLSL++ FV+V     +  KG+ W...\n",
      "KPPYSYIALITMAIQNAPEKKITLNGIYQFIMDRFPFYRENKQGWQNSIRHNLSLNECFVKVPRDDKKPGKGSYW...\n",
      "****Alignment****\n",
      "sequence: pdb|5X07|C Crystal structure of FOXA2 DNA binding domain bound to a full consensus DNA site [Homo sapiens] >pdb|5X07|F Crystal structure of FOXA2 DNA binding domain bound to a full consensus DNA site [Homo sapiens] >pdb|5X07|I Crystal structure of FOXA2 DNA binding domain bound to a full consensus DNA site [Homo sapiens] >pdb|5X07|L Crystal structure of FOXA2 DNA binding domain bound to a full consensus DNA site [Homo sapiens]\n",
      "length: 104\n",
      "e value: 6.35801e-21\n",
      "EVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGA...\n",
      "  +PP++Y SLI  AI +SP K LTL+EIY W   +FP++R+N   W+N++RH+LS +  F++V     +  KG+...\n",
      "HAKPPYSYISLITMAIQQSPNKMLTLSEIYQWIMDLFPFYRQNQQRWQNSIRHSLSFNDCFLKVPRSPDKPGKGS...\n",
      "****Alignment****\n",
      "sequence: pdb|7YZE|A Chain A, Hepatocyte nuclear factor 3-beta [Homo sapiens] >pdb|7YZF|C Chain C, Hepatocyte nuclear factor 3-beta [Homo sapiens]\n",
      "length: 125\n",
      "e value: 8.81924e-21\n",
      "EVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGA...\n",
      "  +PP++Y SLI  AI +SP K LTL+EIY W   +FP++R+N   W+N++RH+LS +  F++V     +  KG+...\n",
      "HAKPPYSYISLITMAIQQSPNKMLTLSEIYQWIMDLFPFYRQNQQRWQNSIRHSLSFNDCFLKVPRSPDKPGKGS...\n",
      "****Alignment****\n",
      "sequence: pdb|1VTN|C Co-Crystal Structure Of The Hnf-3FORK HEAD DNA-Recognition Motif Resembles Histone H5 [Homo sapiens]\n",
      "length: 102\n",
      "e value: 1.96034e-20\n",
      "EVRPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENV-----KGA...\n",
      "  +PP++Y SLI  AI ++P K LTL+EIY W   +FPY+R N   W+N++RH+LS +  FV+V        KG+...\n",
      "HAKPPYSYISLITMAIQQAPGKMLTLSEIYQWIMDLFPYYRENQQRWQNSIRHSLSFNDCFVKVARSPDKPGKGS...\n",
      "****Alignment****\n",
      "sequence: pdb|2UZK|A Crystal structure of the human FOXO3a-DBD bound to DNA [Homo sapiens] >pdb|2UZK|C Crystal structure of the human FOXO3a-DBD bound to DNA [Homo sapiens]\n",
      "length: 97\n",
      "e value: 4.29958e-20\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  SP+K+LTL++IY W  R  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITRAIESSPDKRLTLSQIYEWMVRCVPYFKDKGDSNSSAGWKNSIRHNLSLHSRFMRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|2K86|A Solution Structure of FOXO3a Forkhead domain [Homo sapiens]\n",
      "length: 103\n",
      "e value: 5.73651e-20\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  SP+K+LTL++IY W  R  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITRAIESSPDKRLTLSQIYEWMVRCVPYFKDKGDSNSSAGWKNSIRHNLSLHSRFMRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|7VOV|A Chain A, Forkhead box protein L2 [Homo sapiens] >pdb|7VOV|B Chain B, Forkhead box protein L2 [Homo sapiens]\n",
      "length: 94\n",
      "e value: 7.81954e-20\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGAVW...\n",
      "+PP++Y +LI  AI ES EK+LTL+ IY +    FP++ +N   W+N++RHNLSL++ F++V        KG  W...\n",
      "KPPYSYVALIAMAIRESAEKRLTLSGIYQYIIAKFPFYEKNKKGWQNSIRHNLSLNECFIKVPREGGGERKGNYW...\n",
      "****Alignment****\n",
      "sequence: pdb|7VOU|C Chain C, Forkhead box protein L2 [Homo sapiens]\n",
      "length: 111\n",
      "e value: 1.0007e-19\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV-----ENVKGAVW...\n",
      "+PP++Y +LI  AI ES EK+LTL+ IY +    FP++ +N   W+N++RHNLSL++ F++V        KG  W...\n",
      "KPPYSYVALIAMAIRESAEKRLTLSGIYQYIIAKFPFYEKNKKGWQNSIRHNLSLNECFIKVPREGGGERKGNYW...\n",
      "****Alignment****\n",
      "sequence: pdb|3L2C|A Crystal Structure of the DNA Binding Domain of FOXO4 Bound to DNA [Homo sapiens]\n",
      "length: 110\n",
      "e value: 1.77579e-19\n",
      "TYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      "+YA LI QAI  +PEK+LTL +IY W  R  PYF+      ++A WKN++RHNLSLH  F++V N...\n",
      "SYAELISQAIESAPEKRLTLAQIYEWMVRTVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIKVHN...\n",
      "****Alignment****\n",
      "sequence: pdb|1E17|A Solution structure of the DNA binding domain of the human Forkhead transcription factor AFX (FOXO4) [Homo sapiens]\n",
      "length: 150\n",
      "e value: 2.33819e-19\n",
      "TYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      "+YA LI QAI  +PEK+LTL +IY W  R  PYF+      ++A WKN++RHNLSLH  F++V N...\n",
      "SYAELISQAIESAPEKRLTLAQIYEWMVRTVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIKVHN...\n",
      "****Alignment****\n",
      "sequence: pdb|3CO6|C Crystal Structure of FoxO1 DBD Bound to DBE1 DNA [Homo sapiens]\n",
      "length: 100\n",
      "e value: 3.0393e-19\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  S EK+LTL++IY W  +  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITKAIESSAEKRLTLSQIYEWMVKSVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|5A5U|B Structure of mammalian eIF3 in the context of the 43S preinitiation complex [Oryctolagus cuniculus] >pdb|6FEC|w Chain w, Eukaryotic translation initiation factor 3 subunit B [Homo sapiens]\n",
      "length: 1121\n",
      "e value: 3.92432e-19\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENV-----KGAVW...\n",
      "+PP++Y SLI  AI ++P K LTL+EIY W   +FPY+R+N   W+N++RH+LS +  F++V        KG+ W...\n",
      "KPPYSYISLITMAIQQAPSKMLTLSEIYQWIMDLFPYYRQNQQRWQNSIRHSLSFNDCFLKVPRSPDKPGKGSFW...\n",
      "****Alignment****\n",
      "sequence: pdb|5DUI|A Identification of a new FoxO1 binding site that precludes CREB binding at the glucose-6-phosphatase catalytic subunit gene promoter [Homo sapiens] >pdb|5DUI|B Identification of a new FoxO1 binding site that precludes CREB binding at the glucose-6-phosphatase catalytic subunit gene promoter [Homo sapiens]\n",
      "length: 111\n",
      "e value: 4.26114e-19\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  S EK+LTL++IY W  +  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITKAIESSAEKRLTLSQIYEWMVKSVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|3CO7|C Crystal Structure of FoxO1 DBD Bound to DBE2 DNA [Homo sapiens] >pdb|3CO7|F Crystal Structure of FoxO1 DBD Bound to DBE2 DNA [Homo sapiens] >pdb|3COA|C Crystal Structure of FoxO1 DBD Bound to IRE DNA [Homo sapiens] >pdb|3COA|F Crystal Structure of FoxO1 DBD Bound to IRE DNA [Homo sapiens]\n",
      "length: 117\n",
      "e value: 5.47153e-19\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  S EK+LTL++IY W  +  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITKAIESSAEKRLTLSQIYEWMVKSVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|6QVW|A Solution structure of the free FOXO1 DNA binding domain [Homo sapiens]\n",
      "length: 119\n",
      "e value: 5.50835e-19\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  S EK+LTL++IY W  +  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITKAIESSAEKRLTLSQIYEWMVKSVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|6LBI|C Chain C, Forkhead box protein O1 [Homo sapiens] >pdb|6LBI|D Chain D, Forkhead box protein O1 [Homo sapiens] >pdb|6LBI|G Chain G, Forkhead box protein O1 [Homo sapiens] >pdb|6LBI|H Chain H, Forkhead box protein O1 [Homo sapiens] >pdb|6LBI|K Chain K, Forkhead box protein O1 [Homo sapiens] >pdb|6LBI|L Chain L, Forkhead box protein O1 [Homo sapiens]\n",
      "length: 118\n",
      "e value: 6.25555e-19\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  S EK+LTL++IY W  +  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITKAIESSAEKRLTLSQIYEWMVKSVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|4LG0|A Structure of a ternary FOXO1-ETS1 DNA complex [Homo sapiens]\n",
      "length: 133\n",
      "e value: 1.03047e-18\n",
      "FTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFR-----RNAATWKNAVRHNLSLHKYFVRVEN...\n",
      " +YA LI +AI  S EK+LTL++IY W  +  PYF+      ++A WKN++RHNLSLH  F+RV+N...\n",
      "LSYADLITKAIESSAEKRLTLSQIYEWMVKSVPYFKDKGDSNSSAGWKNSIRHNLSLHSKFIRVQN...\n",
      "****Alignment****\n",
      "sequence: pdb|7FJ2|A Chain A, Forkhead box protein M1 [Homo sapiens] >pdb|7FJ2|B Chain B, Forkhead box protein M1 [Homo sapiens] >pdb|7FJ2|E Chain E, Forkhead box protein M1 [Homo sapiens] >pdb|7FJ2|F Chain F, Forkhead box protein M1 [Homo sapiens]\n",
      "length: 117\n",
      "e value: 2.0183e-18\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNA-ATWKNAVRHNLSLHKYFVRVENVKGAV--WTV...\n",
      "RPP++Y ++I+ AI  +  K++TL +IY W    FPYF+  A   WKN++RHNLSLH  FVR  +  G V  WT+...\n",
      "RPPYSYMAMIQFAINSTERKRMTLKDIYTWIEDHFPYFKHIAKPGWKNSIRHNLSLHDMFVRETSANGKVSFWTI...\n",
      "****Alignment****\n",
      "sequence: pdb|3G73|A Structure of the FOXM1 DNA binding [Homo sapiens] >pdb|3G73|B Structure of the FOXM1 DNA binding [Homo sapiens]\n",
      "length: 142\n",
      "e value: 3.45192e-18\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNA-ATWKNAVRHNLSLHKYFVRVENVKGAV--WTV...\n",
      "RPP++Y ++I+ AI  +  K++TL +IY W    FPYF+  A   WKN++RHNLSLH  FVR  +  G V  WT+...\n",
      "RPPYSYMAMIQFAINSTERKRMTLKDIYTWIEDHFPYFKHIAKPGWKNSIRHNLSLHDMFVRETSANGKVSFWTI...\n",
      "****Alignment****\n",
      "sequence: pdb|5OCN|A Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|B Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|C Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|D Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|E Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|F Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|G Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|5OCN|H Crystal structure of the forkhead domain of human FOXN1 [Homo sapiens] >pdb|6EL8|A Crystal structure of the Forkhead domain of human FOXN1 in complex with DNA [Homo sapiens] >pdb|6EL8|D Crystal structure of the Forkhead domain of human FOXN1 in complex with DNA [Homo sapiens]\n",
      "length: 99\n",
      "e value: 1.01412e-17\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRVENV------KGAV...\n",
      "+P ++Y+ LI  A+  S    L ++EIYN+ T  FPYF+     WKN+VRHNLSL+K F +VEN       KG +...\n",
      "KPIYSYSILIFMALKNSKTGSLPVSEIYNFMTEHFPYFKTAPDGWKNSVRHNLSLNKCFEKVENKSGSSSRKGCL...\n",
      "****Alignment****\n",
      "sequence: pdb|7YZB|A Chain A, Forkhead box protein H1 [Homo sapiens]\n",
      "length: 185\n",
      "e value: 1.0437e-12\n",
      "RPPFTYASLIRQAILESPEKQLTLNEIYNWFTRMFPYFRRNAATWKNAVRHNLSLHKYFVRV------ENVKGAV...\n",
      "+PP+TY ++I   I  +P ++L L +I      +FP+FR +   WK+++RHNLS ++ F +V         KG  ...\n",
      "KPPYTYLAMIALVIQAAPSRRLKLAQIIRQVQAVFPFFREDYEGWKDSIRHNLSSNRCFRKVPKDPAKPQAKGNF...\n"
     ]
    }
   ],
   "source": [
    "for alignment in blast_record.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        print('****Alignment****')\n",
    "        print('sequence:', alignment.title)\n",
    "        print('length:', alignment.length)\n",
    "        print('e value:', hsp.expect)\n",
    "        print(hsp.query[0:75] + '...')\n",
    "        print(hsp.match[0:75] + '...')\n",
    "        print(hsp.sbjct[0:75] + '...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The NCBI Protein accession ID for hemoglobin is 2487155142\n",
      "The NCBI Protein accession ID for insulin is 2487199811\n",
      "The NCBI Protein accession ID for myosin is 2487201619\n",
      "The NCBI Protein accession ID for collagen is 2487200458\n",
      "The NCBI Protein accession ID for immunoglobulin is 2487201646\n",
      "The NCBI Protein accession ID for amylase is 2487158369\n",
      "The NCBI Protein accession ID for trypsin is 2487199163\n",
      "The NCBI Protein accession ID for keratin is 2487199691\n",
      "The NCBI Protein accession ID for fibrinogen is 2487193546\n",
      "The NCBI Protein accession ID for erythropoietin is 2487161332\n",
      "The NCBI Protein accession ID for albumin is 2487193311\n",
      "The NCBI Protein accession ID for elastin is 2487202389\n",
      "The NCBI Protein accession ID for catalase is 2487169621\n",
      "The NCBI Protein accession ID for protein kinase is 2487202646\n",
      "The NCBI Protein accession ID for tubulin is 2487202509\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "    Here I report those prodein with different functions:\n",
    "        \n",
    "        Hemoglobin - transports oxygen in the blood\n",
    "        Insulin - regulates glucose metabolism\n",
    "        Myosin - facilitates muscle contraction\n",
    "        Collagen - provides structural support to tissues\n",
    "        Immunoglobulin - recognizes and neutralizes foreign substances (antigens)\n",
    "        Amylase - breaks down carbohydrates in the digestive system\n",
    "        Trypsin - breaks down proteins in the digestive system\n",
    "        Keratin - forms the structural components of skin, hair, and nails\n",
    "        Fibrinogen - plays a key role in blood clotting\n",
    "        Erythropoietin - stimulates the production of red blood cells\n",
    "        Albumin - maintains the osmotic pressure of blood and transports various molecules\n",
    "        Elastin - provides elasticity to tissues such as skin, arteries, and lungs\n",
    "        Catalase - breaks down hydrogen peroxide in cells\n",
    "        Protein kinase - regulates various cellular processes by modifying other proteins through phosphorylation\n",
    "        Tubulin - forms the structural components of microtubules, which are important for cell division and intracellular transport.\n",
    "\n",
    "\n",
    "        Hemoglobin: NP_000509.1 (beta subunit)\n",
    "        Insulin: NP_000198.1 (preproinsulin)\n",
    "        Actin: NP_001092.2 (actin, alpha cardiac muscle 1)\n",
    "        Collagen: NP_000079.3 (collagen alpha-1(I) chain)\n",
    "        Immunoglobulin: NP_001138834.1 (immunoglobulin heavy constant gamma 1)\n",
    "        Amylase: NP_000681.2 (amylase, alpha 1A)\n",
    "        Trypsin: NP_001239255.1 (trypsinogen 1)\n",
    "        Keratin: NP_006116.2 (keratin, type II cytoskeletal 1)\n",
    "        Fibrinogen: NP_000500.3 (fibrinogen alpha chain)\n",
    "        Erythropoietin: NP_000790.1 (erythropoietin precursor)\n",
    "        Albumin: NP_000468.2 (serum albumin)\n",
    "        Elastin: NP_000088.1 (elastin)\n",
    "        Catalase: NP_001752.1 (catalase)\n",
    "        Protein kinase: NP_001158288.1 (protein kinase C alpha type)\n",
    "        Tubulin: NP_001191.1 (tubulin beta chain)\n",
    "\n",
    "\n",
    "    I will try to retrive sequences which the function is supposed\n",
    "    to be similar, since the similarity in the sequence.\n",
    "        \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def get_protein_id(protein_name, mail: str = 'riccardo.tedoldi@studenti.unitn.it'):\n",
    "    \"\"\" Get the NCBI Protein accession ID for a given protein name.\n",
    "\n",
    "    Args:\n",
    "        protein_name (str): The name of the protein.\n",
    "        mail (str): The email address to use for the Entrez API.\n",
    "\n",
    "    Returns:\n",
    "        str: The NCBI Protein accession ID for the given protein name.\n",
    "\n",
    "    \"\"\"\n",
    "    # Set up the Entrez API\n",
    "    Entrez.email = mail\n",
    "    db = \"protein\"\n",
    "\n",
    "    # Send a search request to the NCBI Protein database\n",
    "    handle = Entrez.esearch(db=db, term=protein_name)\n",
    "\n",
    "    # Parse the search results to extract the NCBI Protein accession ID\n",
    "    record = Entrez.read(handle)\n",
    "    if int(record[\"Count\"]) > 0:\n",
    "        protein_id = record[\"IdList\"][0]\n",
    "        print(f\"The NCBI Protein accession ID for {protein_name} is {protein_id}\")\n",
    "        handle.close()\n",
    "        return protein_id\n",
    "    else:\n",
    "        handle.close()\n",
    "        raise ValueError(f\"No results found for {protein_name}\")\n",
    "\n",
    "# find the accession id of the protein\n",
    "\n",
    "# Define the protein name\n",
    "proteins_name = [\n",
    "\n",
    "    \"hemoglobin\", \n",
    "    \"insulin\",\n",
    "    \"myosin\",\n",
    "    \"collagen\",\n",
    "    \"immunoglobulin\",\n",
    "    \"amylase\",\n",
    "    \"trypsin\",\n",
    "    \"keratin\",\n",
    "    \"fibrinogen\",\n",
    "    \"erythropoietin\",\n",
    "    \"albumin\",\n",
    "    \"elastin\",\n",
    "    \"catalase\",\n",
    "    \"protein kinase\",\n",
    "    \"tubulin\"  \n",
    "\n",
    "]\n",
    "\n",
    "# Get the protein accession ID\n",
    "proteins_accession = [\n",
    "\n",
    "    \"NP_000509.1\", # Hemoglobin\n",
    "    \"NP_000198.1\", # Insulin\n",
    "    \"CAA86293.1\", # Myosin\n",
    "    \"NP_000079.2\", # Collagen\n",
    "    \"CAA41852.1\", # Immunoglobulin\n",
    "    \"NP_000681.2\", # Amylase\n",
    "    \"KGI35851.1\", # Trypsin\n",
    "    \"CAA73943.1\", # Keratin\n",
    "    \"NP_000500.2\", # Fibrinogen\n",
    "    \"NP_000790.1\", # Erythropoietin\n",
    "    \"NP_000468.1\", # Albumin\n",
    "    \"NP_000088.1\", # Elastin\n",
    "    \"NP_001743.1\", # Catalase\n",
    "    \"AAA58363.1\", # Protein kinase\n",
    "    \"NP_001280141.1\" # Tubulin\n",
    "    \n",
    "]\n",
    "\n",
    "# Get the protein ID\n",
    "proteins_id = [get_protein_id(name) for name in proteins_name]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def fetch_protein(proteins_name, protein_id, mail: str = 'riccardo.tedoldi@studenti.unitn.it'):\n",
    "    \"\"\" Fetch a protein sequence from NCBI Protein database and save it in FASTA format\n",
    "\n",
    "    Args:\n",
    "        proteins_name (str): name of the protein\n",
    "        protein_id (str): NCBI Protein accession ID\n",
    "        mail (str): email address to use for the Entrez API\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "\n",
    "    Save the protein sequence in FASTA format in the current working directory\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    # Set up the Entrez API\n",
    "    Entrez.email = mail\n",
    "    db = \"protein\"\n",
    "    path = './proteins/'\n",
    "\n",
    "    #Here, we set up a temporary handle with our downloaded sequence in fasta format\n",
    "    temp = Entrez.efetch(db=db,rettype=\"fasta\",id=protein_id)\n",
    "\n",
    "    #Reading the sequence information as a string in fasta format\n",
    "    seq = SeqIO.read(temp, format=\"fasta\")\n",
    "\n",
    "    #Creating a fasta file to write our downloaded sequence\n",
    "    seq_out = open(f\"{path}{proteins_name}.{protein_id}.fasta\",'w')\n",
    "\n",
    "    #Writing the sequence record in fasta format\n",
    "    SeqIO.write(seq,seq_out,\"fasta\")\n",
    "\n",
    "    #Closing both the temp handle and the FASTA file\n",
    "    temp.close()\n",
    "    seq_out.close()\n",
    "\n",
    "\n",
    "\n",
    "# Fetch the proteins\n",
    "for i in range(len(proteins_name)):\n",
    "    fetch_protein(proteins_name[i], proteins_accession[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Some useful parameters for the BLAST search are:\n",
    "\n",
    "        program blastn, blastp, blastx, tblastn, or tblastx (lower case)\n",
    "\n",
    "        database Which database to search against (e.g. “nr”).\n",
    "\n",
    "        sequence The sequence to search.\n",
    "\n",
    "        ncbi_gi TRUE/FALSE whether to give ‘gi’ identifier.\n",
    "\n",
    "        descriptions Number of descriptions to show. Def 500.\n",
    "\n",
    "        alignments Number of alignments to show. Def 500.\n",
    "\n",
    "        expect An expect value cutoff. Def 10.0.\n",
    "\n",
    "        matrix_name Specify an alt. matrix (PAM30, PAM70, BLOSUM80, BLOSUM45).\n",
    "\n",
    "        filter “none” turns off filtering. Default no filtering\n",
    "\n",
    "        format_type “HTML”, “Text”, “ASN.1”, or “XML”. Def. “XML”.\n",
    "\n",
    "        entrez_query Entrez query to limit Blast search\n",
    "\n",
    "        hitlist_size Number of hits to return. Default 50\n",
    "\n",
    "        megablast TRUE/FALSE whether to use MEga BLAST algorithm (blastn only)\n",
    "\n",
    "        short_query TRUE/FALSE whether to adjust the search parameters for a\n",
    "            short query sequence. Note that this will override manually set \n",
    "            parameters like word size and e value. Turns off when sequence \n",
    "            length is > 30 residues. Default: None.\n",
    "\n",
    "        service plain, psi, phi, rpsblast, megablast (lower case)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# here we perform blastp to retrive similar sequences\n",
    "# that might have the same funtion of the human protein\n",
    "# found. specifically, with blastp we can retrive similar\n",
    "# sequences in the same organism, or in other organisms\n",
    "# this allwos us to discover how the evolution affected\n",
    "# different species\n",
    "\n",
    "import glob\n",
    "import os\n",
    "from Bio.Blast import NCBIWWW\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def blastp(protein_name, db = 'nr'):\n",
    "    \"\"\" Perform a blastp search for a given protein.\n",
    "\n",
    "    Args:\n",
    "        protein_name (str): The name of the protein.\n",
    "        db (str): The database to use for the blastp search.\n",
    "\n",
    "    Returns:\n",
    "        None\n",
    "\n",
    "    Save the blastp results in XML format in the folder \n",
    "    proteins/protein_name/blast_results_protein_name.db.xml\n",
    "\n",
    "    \"\"\"\n",
    "    database = db\n",
    "\n",
    "    path_dir = glob.glob('./proteins/*.fasta')\n",
    "\n",
    "    for path in path_dir:\n",
    "\n",
    "        # get the protein name\n",
    "        protein_name = path.split('.')[1].split('/')[2]\n",
    "\n",
    "        # create a folder for each protein if not already exists\n",
    "        if not os.path.isdir(f'./proteins/{protein_name}'):\n",
    "            os.mkdir(f'./proteins/{protein_name}')\n",
    "\n",
    "        out_blast = f'./proteins/{protein_name}/'\n",
    "\n",
    "        # read the fasta file\n",
    "        seq = SeqIO.read(path, format='fasta')\n",
    "\n",
    "        print(f\"Performing blastp search for {protein_name} in {database} database\")\n",
    "\n",
    "        result_handle = NCBIWWW.qblast(\"blastp\", database, seq.seq)\n",
    "\n",
    "        print(f\"Saving the results for {protein_name} in {database} database\")\n",
    "\n",
    "        # Save the results to a file\n",
    "        with open(f\"{out_blast}blast_results{protein_name}.{database}.xml\", \"w\") as out_handle:\n",
    "            out_handle.write(result_handle.read())\n",
    "\n",
    "        print(f\"Results saved for {protein_name} in {database} database\")\n",
    "        print()\n",
    "            \n",
    "        # Close the handle\n",
    "        result_handle.close()\n",
    "\n",
    "\n",
    "# database = ['pdb', 'nt']\n",
    "\n",
    "blastp(proteins_name)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the results from the files\n",
    "with open(\"./proteins/keratin/keratin-blastp.xml\") as in_handle:\n",
    "    blast_records = NCBIXML.parse(in_handle)\n",
    "    for blast_record in blast_records:\n",
    "        for alignment in blast_record.alignments:\n",
    "            print(f\"Alignment: {alignment.title}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
